File: JAVA/ATM_Design/src/main/java/org/ceo/states/ReadyForTransactionState.java
package org.ceo.states;

import org.ceo.apis.ATMBackendAPI;
import org.ceo.apis.NodeATMBackendAPI;
import org.ceo.enums.ATMState;
import org.ceo.factories.StateFactory;
import org.ceo.models.ATM;

public class ReadyForTransactionState extends ATMStateMachine {
    private final ATM atm;
    private final ATMBackendAPI ATMBackendAPI;
    public ReadyForTransactionState(ATM atm) {
        this.atm = atm;
        this.ATMBackendAPI = new NodeATMBackendAPI();
    }

    @Override
    public int initTransaction() {
        int transactionId = this.ATMBackendAPI.createTransaction();

        if(transactionId == 0) {
            throw new RuntimeException("Transaction could not be created");
        }

        this.atm.changeState(StateFactory.getState(ATMState.READ_CARD_DETAILS_AND_PIN, this.atm));
        return transactionId;
    }

    @Override
    public ATMState getState() {
        return ATMState.READY_FOR_TRANSACTION;
    }
}

File: C++/ATM-System/src/states/ReadyForTransactionState.cpp
#include "../states/ReadyForTransactionState.h"
#include "../states/ReadCardAndPinDetailsState.h"
#include "../models/ATM.h"
#include <iostream>
#include <stdexcept>

int ReadyForTransactionState::initTransaction() {
    int transactionId = 1234;  // fake backend call
    atm->changeState(new ReadCardAndPinDetailsState(atm));
    return transactionId;
}

ATMState ReadyForTransactionState::getState() {
    return ATMState::READY_FOR_TRANSACTION;
}
